### PR TITLE
Move desugaring of various SQL constructs to planner

### PR DESCRIPTION
**Pull request by martint**
This is in preparation for the AST/IR split

## Documentation

(x) No documentation is needed.

## Release notes

(x) No release notes entries required.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 1]
```diff
@@ -1436,6 +1440,38 @@ protected Object visitRow(Row node, Object context)
...
+
+            RowType rowType = anonymous(argumentTypes);
+            ResolvedFunction function = plannerContext.getMetadata()
+                    .resolveFunction(session, QualifiedName.of(FormatFunction.NAME), TypeSignatureProvider.fromTypes(VARCHAR, rowType));
+
+            // Construct a row with arguments [1..n] and invoke the underlying function
+            BlockBuilder rowBuilder = new RowBlockBuilder(argumentTypes, null, 1);
+            BlockBuilder singleRowBlockWriter = rowBuilder.beginBlockEntry();
+            for (int i = 0; i < arguments.size(); ++i) {
+                writeNativeValue(argumentTypes.get(i), singleRowBlockWriter, processWithExceptionHandling(arguments.get(i), context));
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:1466]: Please note that the processed format or any of the arguments might be unresolved. In such case, we should return a (desugared) function call instead of invoking a function.

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 2]
```diff
@@ -1506,6 +1507,37 @@ protected Object visitSubscriptExpression(SubscriptExpression node, Object conte
             return invokeOperator(OperatorType.SUBSCRIPT, types(node.getBase(), node.getIndex()), ImmutableList.of(base, index));
         }
 
+        @Override
+        protected Object visitExtract(Extract node, Object context)
+        {
+            Object value = processWithExceptionHandling(node.getExpression(), context);
+            if (value == null || value instanceof Expression) {
+                return null;
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:1515]: Why return null if the value is an Expression?

**Review by kasiafi — COMMENTED** [review]
I have reviewed all commits except the last two.

Please make sure that you find all sites where Expressions like `CurrentCatalog`, `AtTimeZone`, `Format` etc. are visited past the point of the TranslationMap. They should probably throw `IllegalStateException` or so.

File: core/trino-main/src/main/java/io/trino/sql/planner/DomainTranslator.java [hunk 3]
```diff
@@ -815,7 +818,8 @@ private static Domain extractEquatableDomain(ComparisonExpression.Operator compa
             return switch (comparisonOperator) {
                 case EQUAL -> Domain.create(complementIfNecessary(ValueSet.of(type, value), complement), false);
                 case NOT_EQUAL -> Domain.create(complementIfNecessary(ValueSet.of(type, value).complement(), complement), false);
-                case IS_DISTINCT_FROM -> complementIfNecessary(Domain.create(ValueSet.of(type, value).complement(), true), complement); // Need to potential complement the whole domain for IS_DISTINCT_FROM since it is null-aware
+                case IS_DISTINCT_FROM ->
+                        complementIfNecessary(Domain.create(ValueSet.of(type, value).complement(), true), complement); // Need to potential complement the whole domain for IS_DISTINCT_FROM since it is null-aware
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/DomainTranslator.java:822]: Unrelated formatting change?

**Review by kasiafi — COMMENTED** [review]
I have reviewed the remaining commits.

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java (hunk 2 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:1515]: Good catch. That's wrong.

**Comment by martint** [comment]
> Please make sure that you find all sites where Expressions like CurrentCatalog, AtTimeZone, Format etc. are visited past the point of the TranslationMap.

Which places are you thinking of? We already have a checker to ensure none of those leak past the optimizer.

**Comment by martint** [comment]
@kasiafi, updated

**Review by kasiafi — COMMENTED** [review]
> Which places are you thinking of? We already have a checker to ensure none of those leak past the optimizer.

I was thinking of the optimizer. In the normal flow of the query processing, none of those should appear there. We could't exclude them in the ExpressionInterpreter, as it is accessed both from the optimizer and analyzer.

I checked some of the expressions and haven't seen them explicitly visited in the optimizer, so my concern is probably invalid.

> We already have a checker to ensure none of those leak past the optimizer.

I found that `Format` or `CurrentTime` are not caught by the `SugarFreeChecker`.

**Comment by kasiafi** [comment]
Adding to the comments on non-desugared expression returned by the `ExpressionInterpreter`, `Like` also behaves this way, an it is preexisting.

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 4]
```diff
@@ -1458,6 +1458,17 @@ protected Object visitFormat(Format node, Object context)
                     .map(this::type)
                     .collect(toImmutableList());
 
+            List<Object> processedArguments = arguments.stream()
+                    .map(argument -> processWithExceptionHandling(argument, context))
+                    .collect(toList());
+
+            if (hasUnresolvedValue(processedArguments)) {
+                return new Format(ImmutableList.<Expression>builder()
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:1467]: It's ok. There's no benefit in returning the desugared expressions since they will get translated during planning. In fact, it means we're duplicating behavior in two places.

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 5]
```diff
@@ -1335,14 +1335,18 @@ protected Object visitAtTimeZone(AtTimeZone node, Object context)
             Type valueType = type(node.getValue());
             Type timeZoneType = type(node.getTimeZone());
 
+            if (value instanceof Expression || timeZone instanceof Expression) {
+                return new AtTimeZone(toExpression(value, valueType), toExpression(timeZone, timeZoneType));
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:1346]: See other comment.

File: core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java [hunk 6]
```diff
@@ -1359,6 +1522,41 @@ protected Object visitSubscriptExpression(SubscriptExpression node, Object conte
...
+        @Override
+        protected Object visitExtract(Extract node, Object context)
+        {
+            Object value = processWithExceptionHandling(node.getExpression(), context);
+            if (value == null) {
+                return null;
+            }
+
+            if (value instanceof Expression) {
+                return new Extract(toExpression(value, type(node)), node.getField());
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java:1534]: See other comment

**Review by kasiafi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java [hunk 7]
```diff
@@ -646,6 +650,42 @@ public Expression rewriteTryExpression(TryExpression node, Void context, Express
...
+                            .build();
+                }
+                else {
+                    patternCall = FunctionCallBuilder.resolve(session, plannerContext.getMetadata())
+                            .setName(QualifiedName.of(LIKE_PATTERN_FUNCTION_NAME))
+                            .addArgument(analysis.getType(node.getPattern()), pattern)
+                            .build();
+                }
+
+                FunctionCall call = FunctionCallBuilder.resolve(session, plannerContext.getMetadata())
```
Comment (assaf2) [inline core/trino-main/src/main/java/io/trino/sql/planner/TranslationMap.java:680]: @martint this change causes the following impact:
Since now we transform `LikePredicate` into a `FunctionCall`, `io.trino.sql.tree.ExpressionRewriter#rewriteLikePredicate` is not called anymore and therefore, for example, doesn't turn `char(1) LIKE 'a'` into a `ComparisonExpression`. `DomainTranslator` doesn't process `like` of char type so now we push down such filter as a `ConnectorExpression` instead of `TupleDomain`
